### PR TITLE
Upgrade to reflection-docblock 3

### PR DESCRIPTION
**Pull request by dunglas**
Tests will be green when phpDocumentor/TypeResolver#16 and phpDocumentor/ReflectionDocBlock#72 will be merged.

### CODE REVIEW

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php [hunk 1]
```diff
@@ -44,23 +57,30 @@ public function apply(ClassNode $node)
...
+            $phpdoc = $this->docBlockFactory->create($reflectionClass, $this->contextFactory->createFromReflector($reflectionClass));
+        } catch (\InvalidArgumentException $e) {
+            // No DocBlock
+        }
 
-        $tagList = $phpdoc->getTagsByName('method');
+        /**
+         * @var Method[] $tagList
+         */
+        $tagList = isset($phpdoc) ? $phpdoc->getTagsByName('method') : array();
```
Comment (stof) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: I don't like the idea of dealing with codepath where the variable may be defined or no. The code should probably be refactored by moving this inside the try/catch

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php [hunk 2]
```diff
@@ -44,23 +57,30 @@ public function apply(ClassNode $node)
...
-        $phpdoc = new DocBlock($reflectionClass->getDocComment());
+        try {
+            $phpdoc = $this->docBlockFactory->create($reflectionClass, $this->contextFactory->createFromReflector($reflectionClass));
+        } catch (\InvalidArgumentException $e) {
+            // No DocBlock
+        }
 
-        $tagList = $phpdoc->getTagsByName('method');
+        /**
+         * @var Method[] $tagList
```
Comment (sroze) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: If you really want it, I would use a single line comment instead.

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php [hunk 3]
```diff
@@ -19,9 +22,19 @@
...
  *
  * @author Thomas Tourlourat <thomas@tourlourat.com>
+ * @author Kévin Dunglas <dunglas@gmail.com>
  */
 class MagicCallPatch implements ClassPatchInterface
 {
+    private $docBlockFactory;
+    private $contextFactory;
+
+    public function __construct()
```
Comment (sroze) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: Can you make these dependencies as optionally injectable?

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php [hunk 4]
```diff
@@ -44,23 +57,30 @@ public function apply(ClassNode $node)
...
+                $tagList = array_merge($tagList, $phpdoc->getTagsByName('method'));
+            } catch (\InvalidArgumentException $e) {
+                // No DocBlock
+            }
         }
 
         foreach($tagList as $tag) {
             $methodName = $tag->getMethodName();
 
-            if (empty($methodName)) {
```
Comment (sroze) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: You're sure that with the version 3.0, it won't return empty method names? You're reverting a 12 days old PR (#260) here.

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php (hunk 3 above)
Comment (stof) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: I don't think it is worth it. You generally don't instantiate MagicCallPatch yourselves anyway.

And exposing them would mean that we then have to provide backward compatibility with people injecting us a v3 DocBlockFactory.
We can still change this later if there is a use case to customize the parsing of docblocks

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php (hunk 4 above)
Comment (stof) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: Well, the testsuite will tell us

**Comment by stof** [comment]
This PR means that the min requirement of Prophecy is now PHP 5.5 though

File: src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php (hunk 3 above)
Comment (sroze) [inline src/Prophecy/Doubler/ClassPatch/MagicCallPatch.php]: Yep, this is fair enough we can open the doors after.

**Comment by aik099** [comment]
So Symfony wants Reflection v3, but Prophecy wants Reflection v2?

**Comment by sroze** [comment]
We need to be able to support the two versions, else it will again cause a lot of problems. Based on the small amount of changes that are required, that's fair to abstract it and uses a basic `class_exists` to create the correct adapter here IMO.

**Comment by dunglas** [comment]
I fully agree, will do it ASAP.

**Comment by dunglas** [comment]
Closing in favor of #264

**Pull request closed without merging**